lightblue/src/main/java/org/esbtools/eventhandler/lightblue/LightblueDocumentEventRepository.java

```diff
@@ -92,30 +101,77 @@ public void addNewDocumentEvents(Collection<? extends DocumentEvent> documentEve
+            throws Exception {
+        String[] typesToProcess = getSupportedAndEnabledEventTypes();
+        Integer documentEventsBatchSize = config.getDocumentEventsBatchSize();
+
+        if (typesToProcess.length == 0 || documentEventsBatchSize == null ||
+                documentEventsBatchSize == 0) {
+            logger.info("Not retrieving any document events because either there are no enabled " +
+                    "or supported types to process or documentEventBatchSize is 0. Supported " +
+                    "types are {}. Of those, enabled types are {}. " +
+                    "Document event batch size is {}.",
+                    supportedTypes, Arrays.toString(typesToProcess), documentEventsBatchSize);
+            return Collections.emptyList();
+        }
 
-            DocumentEventEntity[] documentEventEntities = lightblue
-                    .data(FindRequests.priorityDocumentEventsForEntitiesUpTo(entities, documentEventBatchSize))
-                    .parseProcessed(DocumentEventEntity[].class);
+        if (maxEvents == 0) {
+            return Collections.emptyList();
+        }
 
-            if (documentEventEntities.length == 0) {
-                return Collections.emptyList();
-            }
+        DocumentEventEntity[] documentEventEntities = lightblue
+                .data(FindRequests.priorityDocumentEventsForTypesUpTo(
+                        typesToProcess, documentEventsBatchSize,
+                        ZonedDateTime.now(clock).minus(processingTimeout)))
+                .parseProcessed(DocumentEventEntity[].class);
+
+        if (documentEventEntities.length == 0) {
+            return Collections.emptyList();
+        }
+
+        try (LockedResources<SharedIdentityEvents> eventLocks =
+                     SharedIdentityEvents.parseAndOptimizeLockableDocumentEventEntities(
+                             maxEvents,
+                             documentEventEntities,
+                             new BulkLightblueRequester(lightblue),
+                             documentEventFactoriesByType,
+                             lockStrategy,
+                             clock)) {
+
+            return persistNewEntitiesAndStatusUpdatesToExisting(eventLocks);
+        }
+    }
 
-            List<DocumentEventEntity> entitiesToUpdate = new ArrayList<>();
-            BulkLightblueRequester requester = new BulkLightblueRequester(lightblue);
+    /**
+     * {@inheritDoc}
+     *
+     * <p>N.B. This implementation currently works by simply checking if the known in memory
+     * timeouts of provided events fall within some threshold. It could be updated to also timestamp
+     * the persisted events again, but to be safe we would probably want to also lock them to do
+     * this update. For now, that complexity is probably not worth it.
+     */
+    @Override
+    public Collection<? extends DocumentEvent> checkExpired(Collection<? extends DocumentEvent> events) {
```

Comment:
It checks only timeouts, no locks. At this point there are no locks which we care about, we don't keep the identity lock for the entire duration of processing. If we need to do that, it gets a little complicated. I would have to keep a pointer to the locked resource with each event with how the API is currently. Another alternative would be to update the persisted timestamps at this point, but to do that we would want a lock to prevent reads on these processing events while we're updating the timestamps.

I think it's just simpler to use the timeout in memory for now which is not perfect but should be good enough to keep away duplicates and still handle failure.